src/gui/torrentcreatordlg.cpp

```diff
@@ -28,254 +28,239 @@
-#include "guiiconprovider.h"
 #include "base/bittorrent/session.h"
-#include "base/bittorrent/torrentinfo.h"
 #include "base/bittorrent/torrentcreatorthread.h"
+#include "base/bittorrent/torrentinfo.h"
+#include "base/global.h"
+#include "base/utils/fs.h"
 
-const uint NB_PIECES_MIN = 1200;
-const uint NB_PIECES_MAX = 2200;
+#include "ui_torrentcreatordlg.h"
 
-TorrentCreatorDlg::TorrentCreatorDlg(QWidget *parent)
+#define SETTINGS_KEY(name) "TorrentCreator/" name
+
+TorrentCreatorDlg::TorrentCreatorDlg(QWidget *parent, const QString &defaultPath)
     : QDialog(parent)
-    , m_creatorThread(0)
+    , ui(new Ui::TorrentCreatorDlg)
+    , m_creatorThread(new BitTorrent::TorrentCreatorThread(this))
+    , m_storeDialogSize(SETTINGS_KEY("Dimension"))
+    , m_storePieceSize(SETTINGS_KEY("PieceSize"))
+    , m_storePrivateTorrent(SETTINGS_KEY("PrivateTorrent"))
+    , m_storeStartSeeding(SETTINGS_KEY("StartSeeding"))
+    , m_storeIgnoreRatio(SETTINGS_KEY("IgnoreRatio"))
+    , m_storeLastAddPath(SETTINGS_KEY("LastAddPath"), QDir::homePath())
+    , m_storeTrackerList(SETTINGS_KEY("TrackerList"))
+    , m_storeWebSeedList(SETTINGS_KEY("WebSeedList"))
+    , m_storeComments(SETTINGS_KEY("Comments"))
+    , m_storeLastSavePath(SETTINGS_KEY("LastSavePath"), QDir::homePath())
 {
-    setupUi(this);
-    // Icons
-    addFile_button->setIcon(GuiIconProvider::instance()->getIcon("document-new"));
-    addFolder_button->setIcon(GuiIconProvider::instance()->getIcon("folder-new"));
-    createButton->setIcon(GuiIconProvider::instance()->getIcon("document-save"));
-    cancelButton->setIcon(GuiIconProvider::instance()->getIcon("dialog-cancel"));
-
+    ui->setupUi(this);
     setAttribute(Qt::WA_DeleteOnClose);
-    setModal(true);
-    showProgressBar(false);
-    loadTrackerList();
-    // Piece sizes
-    m_pieceSizes << 16 << 32 << 64 << 128 << 256 << 512 << 1024 << 2048 << 4096 << 8192 << 16384;
+    setModal(false);
+
+    ui->buttonBox->button(QDialogButtonBox::Ok)->setText(tr("Create Torrent"));
+
+    connect(ui->addFile_button, SIGNAL(clicked(bool)), SLOT(onAddFileButtonClicked()));
+    connect(ui->addFolder_button, SIGNAL(clicked(bool)), SLOT(onAddFolderButtonClicked()));
+    connect(ui->buttonBox, SIGNAL(accepted()), SLOT(onCreateButtonClicked()));
+
+    connect(m_creatorThread, SIGNAL(creationSuccess(QString, QString)), this, SLOT(handleCreationSuccess(QString, QString)));
+    connect(m_creatorThread, SIGNAL(creationFailure(QString)), this, SLOT(handleCreationFailure(QString)));
+    connect(m_creatorThread, SIGNAL(updateProgress(int)), this, SLOT(updateProgressBar(int)));
+
     loadSettings();
+    updateInputPath(defaultPath);
+
     show();
 }
 
 TorrentCreatorDlg::~TorrentCreatorDlg()
 {
-    if (m_creatorThread)
+    saveSettings();
+
+    // End torrent creation thread
+    if (m_creatorThread) {
+        if (m_creatorThread->isRunning()) {
+            m_creatorThread->abortCreation();
+            m_creatorThread->terminate();
```

Comment:
If @glassez did not mean that there is no way to terminate another thread in general case, I would like to point this out. And If it is not joinable after calling `m_creatorThread->abortCreation()`, there is no much one can do with a hanged thread.